readme.md

# tolkien <img src="man/figures/logo.png" align="right" width="120" />


[![cran version](http://www.r-pkg.org/badges/version/tolkien)](https://cran.r-project.org/package=tolkien)
[![rstudio mirror per-month downloads](http://cranlogs.r-pkg.org/badges/tolkien)](https://github.com/metacran/cranlogs.app)
[![rstudio mirror total downloads](http://cranlogs.r-pkg.org/badges/grand-total/tolkien?color=yellowgreen)](https://github.com/metacran/cranlogs.app)

This package provides the first round of palettes derived from the *Tolkien* universe.

They have been chosen manually, taking into account its consistency with all the existing branding of the franchise, but its suitability for data visualisation. 

> It's the job that's never started as takes longest to finish.

The colour palette should be beautiful, useful for plotting data and shoulr relate to desired style; in this case, should relate to the Tolkien legendarium. Some of the colours might change in future versions, in order to find this balance between suitability for plotting and relatable to the Tolkien legendarium.

<center><img src="https://raw.githubusercontent.com/aljrico/tolkien/master/readme_raw_files/examples/scales_houses.png" ></center>




Installation
-------------

Just copy and execute this bunch of code and you'll have the last version of the package installed:


```r
install.packages("tolkien")
```

And you can now use it:


```r
library(tolkien)
```

Usage
-----

The default colour scale of the package is the one named **Sauron**. If you prefer to choose another one, you'll need to specify which house you want the palette from.

You can get started using base R plot functions, and the `tolkien()` function. Its first argument `n` lets you set the number of colours to be mapped. This way you can set different resolutions. Let's say that you want a palette inspired by **Sauron**, then just type `option = "Sauron"`. 


```r
pal <- tolkien(n = 8, house = "Sauron")
image(volcano, col = pal)

pal <- tolkien(n = 128, house = "Sauron")
image(volcano, col = pal)
```

<center><img src="https://raw.githubusercontent.com/aljrico/tolkien/master/readme_raw_files/examples/sauron_volcano.png" ></center>

## ggplot2

Of course, this package has specific functions to behave seamlessly with the best data visiualisation library available. 
The package contains colour scale functions for **ggplot2** plots: `scale_colour_tolkien()` and `scale_fill_tolkien()`. 


Here is a made up example using the colours from the house of **Hufflepuff**,


```r
library(ggplot2)
ggplot(data.frame(x = rnorm(1e4), y = rnorm(1e4)), aes(x = x, y = y)) +
  geom_hex() + 
	coord_fixed() +
  scale_fill_tolkien(house = "Sauron") + 
	theme_bw()
```

<center><img src="https://raw.githubusercontent.com/aljrico/aljrico.github.io/master/_posts/images/unnamed-chunk-6-1.png" ></center>


or more made-up heatmaps:

<center><img src="https://raw.githubusercontent.com/aljrico/aljrico.github.io/master/_posts/images/unnamed-chunk-8-1.png" ></center>


Using the same function we can also plot these cloropleth maps of U.S. unemployment:

<center><img src="https://raw.githubusercontent.com/aljrico/aljrico.github.io/master/_posts/images/ggplot2-1.png" ></center>

<center><img src="https://raw.githubusercontent.com/aljrico/aljrico.github.io/master/_posts/images/ggplot2-2.png" ></center>


### Discrete Scales

But what if you want discrete scales? These functions also can be used for discrete scales with the argument `discrete = TRUE`. This argument, when TRUE, sets a finite number of sufficiently spaced colours within the selected palette to plot your data.


```r
ggplot(diamonds, aes(factor(color), fill=factor(cut))) +
	geom_bar(colour = "black") +
	scale_fill_tolkien(discrete = TRUE, option = "Sauron", name = "Cut") +
	ylab("") +
	xlab("Colour") +
	coord_flip()
```

<center><img src="https://raw.githubusercontent.com/aljrico/tolkien/master/readme_raw_files/examples/ronweasley_bar.png" ></center>

You can also use discrete scales by adding `_d()` at the end of it instead, like `scale_fill_tolkien_d()`.


```r
dsamp <- diamonds[sample(nrow(diamonds), 1000), ]
ggplot(dsamp, aes(carat, price)) +
	geom_point(aes(colour = clarity)) +
	scale_colour_tolkien_d(option = "Sauron", name = "Clarity") +
	xlab("Carat") +
	ylab("Price")
```

<center><img src="https://raw.githubusercontent.com/aljrico/tolkien/master/readme_raw_files/examples/Sauron_scatter.png" ></center>


Don't forget to try them all.

<center><img src="https://raw.githubusercontent.com/aljrico/tolkien/master/readme_raw_files/examples/mischief_always_tile.png" ></center>



# Palettes

```r
option = "Sauron"
```
<center><img src="https://raw.githubusercontent.com/aljrico/tolkien/master/readme_raw_files/palettes/sauron.png" width = "75%""; height="75%"></center>

---



NEWS.MD
# tolkien 0.1.0

generate_data.R
library(tidyverse)
library(data.table)

map <- list()

# Define colours
sauron <- c("#180010", "#4E2729", "#df7b74", "#E8A577", "#fcd882")

options <- list(sauron = sauron
                )



# Expand palette to accept contiuous scales or longer discrete scales
complete_palette <- function(option, n = 3e3){
	complete_col <- c()
	for(i in 1:(length(option)-1)){
		cols <- colorRampPalette(c(option[i], option[i+1]))
		complete_col <- c(complete_col, cols(n))
	}
	return(complete_col)
}

# Build DF map
make_map <- function(option){
	options[[option]] %>%
		complete_palette() %>%
		grDevices::col2rgb() %>%
		t() %>%
		as.data.frame() %>%
		dplyr::rename(V1 = red) %>%
		dplyr::rename(V2 = green) %>%
		dplyr::rename(V3 = blue) %>%
		dplyr::mutate(option = option)
}

for(h in names(options)){
	df <- make_map(h)
	map <- rbind(map,df)
}


tolkien.map <- map
save(tolkien.map, file = "data/tolkien.map.rda", ascii = FALSE, compress = 'xz')
usethis::use_data(tolkien.map, internal = TRUE, overwrite = TRUE)



R/tolkien_palette.R
#' Original 'Tolkien' colour map
#'
#' A dataset containing some colour palettes inspired on the Tolkien Universe
#'
#'
#'@format A data frame containing all the colours used in the palette:
#'\itemize{
#'   \item V1: Red value
#'   \item V2: Green value
#'   \item V3: Blue value
#'   \item option: It is intended to be a general option for choosing the specific colour palette.
#'}
"tolkien.map"



#' Available Palettes.
#'
#' This function outpus a vector containing the names of all the available palettes in the 'tolkien' package.
#'
#' @return \code{tolkien_palettes} returns a character vector with the names of the palettes available to use.
#'
#' @author Alejandro Jiménez Rico \email{aljrico@@gmail.com}
#'
#'
#' @examples
#' tolkien_palettes()
#'
#' @rdname tolkien_palettes
#' @export

tolkien_palettes <- function(){
	return(unique(tolkien.map$option))
}

#'
#'
#'
#'
#' Tolkien Colour Map.
#'
#' This function creates a vector of \code{n} equally spaced colors along the
#' 'tolkien colour map' of your selection
#'
#' @param n The number of colors (\eqn{\ge 1}) to be in the palette.
#'
#' @param alpha	The alpha transparency, a number in [0,1], see argument alpha in
#' \code{\link[grDevices]{hsv}}.
#'
#' @param begin The (corrected) hue in [0,1] at which the tolkien colormap begins.
#'
#' @param end The (corrected) hue in [0,1] at which the tolkien colormap ends.
#'
#' @param direction Sets the order of colors in the scale. If 1, the default, colors
#' are ordered from darkest to lightest. If -1, the order of colors is reversed.
#'
#' @param option A character string indicating the colourmap from a option to use.
#'
#' @param house A character string indicating the colourmap from a option to use. This parameter is deprectaed, 'option' should be used instead.
#'
#' @return \code{tolkien} returns a character vector, \code{cv}, of color hex
#' codes. This can be used either to create a user-defined color palette for
#' subsequent graphics by \code{palette(cv)}, a \code{col =} specification in
#' graphics functions or in \code{par}.
#'
#' @author Alejandro Jiménez Rico \email{aljrico@@gmail.com}, \href{https://aljrico.github.io}{Personal Blog}
#'
#' @details
#'
#' \if{html}{Here are the color scales:
#'
#'   \out{<div style="text-align: center">}\figure{hogwarts-scales.png}{houses: style="width:750px;max-width:90\%;"}\out{</div>}
#'
#'   }
#' \if{latex}{Here are the color scales:
#'
#'   \out{\begin{center}}\figure{hogwarts-scales.png}\out{\end{center}}
#'   }
#'
#'
#' Semi-transparent colors (\eqn{0 < alpha < 1}) are supported only on some
#' devices: see \code{\link[grDevices]{rgb}}.
#'
#' @examples
#' library(ggplot2)
#' library(hexbin)
#'
#' dat <- data.frame(x = rnorm(1e4), y = rnorm(1e4))
#' ggplot(dat, aes(x = x, y = y)) +
#'   geom_hex() +
#'   coord_fixed() +
#'   scale_fill_gradientn(colours = tolkien(128, option = 'Sauron'))
#'
#' pal <- tolkien(256, option = "Sauron")
#' image(volcano, col = pal)
#'
#' @rdname tolkien
#' @export
#'
tolkien <- function(n, alpha = 1, begin = 0, end = 1, direction = 1, option = 'Sauron', house = NULL) {

	if(!is.null(house)) option <- house

	option <- tolower(option)
	option <- gsub(" ", "", option, fixed = TRUE)
	option <- gsub("\\_", "", option, fixed = FALSE)

	if (begin < 0 | begin > 1 | end < 0 | end > 1) {
		stop("begin and end must be in [0,1]")
	}

	if (abs(direction) != 1) {
		stop("direction must be 1 or -1")
	}

	if (direction == -1) {
		tmp <- begin
		begin <- end
		end <- tmp
	}

	colnames(tolkien.map) <- c("R", "G", "B", "option")

	map <- tolkien.map[tolkien.map$option == option, ]

	map_cols <- grDevices::rgb(map$R, map$G, map$B, maxColorValue = 255)
	fn_cols <- grDevices::colorRamp(map_cols, space = "Lab", interpolate = "spline")
	cols <- fn_cols(seq(begin, end, length.out = n)) / 255
	grDevices::rgb(cols[, 1], cols[, 2], cols[, 3], alpha = alpha)
}



#' @rdname tolkien
#'
#' @export
tolkien_pal <- function(alpha = 1, begin = 0, end = 1, direction = 1, option = 'Sauron', house = NULL) {

	if(!is.null(house)) option <- house
	option <- tolower(option)
	option <- gsub(" ", "", option, fixed = TRUE)
	option <- gsub("\\_", "", option, fixed = FALSE)

	function(n) {
		tolkien(n, alpha, begin, end, direction, option)
	}
}


#' @rdname scale_tolkien
#'
#' @importFrom ggplot2 scale_fill_gradientn scale_color_gradientn discrete_scale
#'
#' @export
scale_color_tolkien <- function(option = 'Sauron', ..., alpha = 1, begin = 0, end = 1, direction = 1,
													 discrete = FALSE, house = NULL) {

	if(!is.null(house)) option <- house
	option <- tolower(option)
	option <- gsub(" ", "", option, fixed = TRUE)
	option <- gsub("\\_", "", option, fixed = FALSE)

	if (discrete) {
		discrete_scale("colour", "tolkien", tolkien_pal(alpha, begin, end, direction, option), ...)
	} else {
		scale_color_gradientn(colours = tolkien(256, alpha, begin, end, direction, option), ...)
	}
}

#' @rdname scale_tolkien
#' @aliases scale_color_tolkien
#' @importFrom ggplot2 discrete_scale
#' @export
scale_colour_tolkien <- scale_color_tolkien

#' @rdname scale_tolkien
#' @aliases scale_color_tolkien
#' @export
scale_colour_tolkien_d <- function(option = 'Sauron', ..., alpha = 1, begin = 0, end = 1,
															direction = 1) {
	discrete_scale("colour", "tolkien", tolkien_pal(alpha, begin, end, direction, option), ...)
}

#' @rdname scale_tolkien
#' @aliases scale_color_tolkien
#' @export
scale_color_tolkien_d <- scale_colour_tolkien_d


#' @rdname scale_tolkien
#' @aliases scale_fill_tolkien
#' @importFrom ggplot2 discrete_scale
#' @export
scale_fill_tolkien_d <- function(option = 'Sauron', ..., alpha = 1, begin = 0, end = 1,
														direction = 1) {
	discrete_scale("fill", "tolkien", tolkien_pal(alpha, begin, end, direction, option), ...)
}


#' @rdname tolkien
#' @aliases tolkien
#' @export
tolkien <- tolkien

#' Talkien colour scales
#'
#' Uses the Tolkien color scale.
#'
#' For \code{discrete == FALSE} (the default) all other arguments are as to
#' \link[ggplot2]{scale_fill_gradientn} or \link[ggplot2]{scale_color_gradientn}.
#' Otherwise the function will return a \code{discrete_scale} with the plot-computed
#' number of colors.
#'
#'
#' @param ... parameters to \code{discrete_scale} or \code{scale_fill_gradientn}
#'
#' @param alpha pass through parameter to \code{tolkien}
#'
#' @param begin The (corrected) hue in [0,1] at which the tolkien colormap begins.
#'
#' @param end The (corrected) hue in [0,1] at which the tolkien colormap ends.
#'
#' @param direction Sets the order of colors in the scale. If 1, the default, colors
#' are as output by \code{tolkien_pal}. If -1, the order of colors is reversed.
#'
#' @param discrete generate a discrete palette? (default: \code{FALSE} - generate continuous palette)
#'
#' @param option A character string indicating the colourmap  to use.
#'
#' @param house A character string indicating the colourmap from a option to use. This parameter is deprectaed, 'option' should be used instead.
#'
#' @rdname scale_tolkien
#'
#' @author Alejandro Jiménez Rico \email{aljrico@@gmail.com}
#'
#' @importFrom ggplot2 scale_fill_gradientn scale_color_gradientn discrete_scale
#'
#' @importFrom gridExtra grid.arrange
#'
#' @examples
#' library(ggplot2)
#'
#'
#'
#' ggplot(mtcars, aes(factor(cyl), fill=factor(vs))) +
#' geom_bar() +
#' scale_fill_tolkien(discrete = TRUE, option = "Sauron")
#'
#' ggplot(mtcars, aes(factor(gear), fill=factor(carb))) +
#' geom_bar() +
#' scale_fill_tolkien(discrete = TRUE, option = "Sauron")
#'
#' ggplot(mtcars, aes(x = mpg, y = disp, colour = hp)) +
#' geom_point(size = 2) +
#' scale_colour_tolkien(option = "Sauron")
#'
#'
#'
#'
#' @export
scale_fill_tolkien <- function(option = 'Sauron', ..., alpha = 1, begin = 0, end = 1, direction = 1,
													discrete = FALSE, house = NULL) {

	if(!is.null(house)) option <- house
	option <- tolower(option)
	option <- gsub(" ", "", option, fixed = TRUE)
	option <- gsub("\\_", "", option, fixed = FALSE)

	if (discrete) {
		discrete_scale("fill", "tolkien", tolkien_pal(alpha, begin, end, direction, option), ...)
	} else {
		scale_fill_gradientn(colours = tolkien(256, alpha, begin, end, direction, option), ...)
	}
}

readme_raw_files/example4.R
library(tidyverse)
devtools::install_github("aljrico/tolkien")
library(tolkien)
library(gridExtra)


# ronweasley_bar -----------------------------------------------------

gg <- ggplot(diamonds, aes(factor(color), fill=factor(cut))) +
	geom_bar(colour = "black") +
	scale_fill_tolkien_d(option = "Sauron", name = "Cut") +
	ylab("") +
	xlab("Colour") +
	coord_flip()

ggsave("examples/ronweasley_bar.png", gg, width = 300, height = 110, units = "mm", device = "png", dpi = "retina")



# lunalovegood_scatter ----------------------------------------------------


dsamp <- diamonds[sample(nrow(diamonds), 1000), ]
gg <- ggplot(dsamp, aes(carat, price)) +
	geom_point(aes(colour = clarity)) +
	scale_colour_tolkien(discrete = TRUE, option = "Sauron", name = "Clarity") +
	xlab("Carat") +
	ylab("Price")

ggsave("examples/lunalovegood_scatter.png", gg, width = 300, height = 110, units = "mm", device = "png", dpi = "retina")



# sprout_bar --------------------------------------------------------------


gg <- ggplot(diamonds, aes(x = price, fill = cut)) +
	geom_histogram(position = "dodge", binwidth = 1000) +
	scale_x_continuous(limits = c(4000,12000)) +
	xlab("Price") +
	ylab("") +
	scale_fill_tolkien_d(option = "Sauron",  name = "Cut")

ggsave("examples/sprout_bar.png", gg, width = 300, height = 110, units = "mm", device = "png", dpi = "retina")



# positive/negative -------------------------------------------------------

df <- data.frame(ID = c(1:10),Diff = c(-5:4))
df$colour <- ifelse(df$Diff < 0, "firebrick1","steelblue")
df$hjust <- ifelse(df$Diff > 0, 1.3, -0.3)
df$colour <- ifelse(df$Diff < 0, "negative","positive")
gg <- ggplot(df, aes(ID, Diff, fill = colour))+
	geom_bar(stat = "identity", position = "identity", colour = "black", size = 0.4) +
	scale_fill_tolkien_d(option = "Sauron", direction = -1, name = "", labels = c("Negative", "Positive"))

ggsave("examples/newtscamander_posneg.png", gg, width = 300, height = 110, units = "mm", device = "png", dpi = "retina")



# mischief_tile -----------------------------------------------------------


gg1 <- ggplot(faithfuld) +
	geom_tile(aes(waiting, eruptions, fill = density)) +
	xlab("Waiting") +
	ylab("Eruptions") +
	scale_fill_tolkien(option = "Sauron", name = "Density") +
	theme_minimal() +
	ggtitle("Sauron")

gg2 <- ggplot(faithfuld) +
	geom_tile(aes(waiting, eruptions, fill = density)) +
	xlab("Waiting") +
	ylab("Eruptions") +
	scale_fill_tolkien(option = "Sauron", name = "Density") +
	theme_minimal() +
	ggtitle("Sauron")

gg <- grid.arrange(gg1,gg2, ncol = 2)

ggsave("examples/mischief_always_tile.png", gg, width = 300, height = 110, units = "mm", device = "png", dpi = "retina")


# hufflepuff_histogram ----------------------------------------------------

gg <- ggplot(mpg, aes(displ)) +
	geom_histogram(aes(fill = class),
								 binwidth = .1,
								 col = "black",
								 size = 0.1) +
	labs(title = "Discrete Scales",
			 subtitle = "Engine Displacement across Vehicle Classes") +
	geom_histogram(aes(fill = class),
								 bins = 5,
								 col = "black",
								 size = 0.1) +   # change number of bins
	labs(title="Histogram with Fixed Bins",
			 subtitle="Engine Displacement across Vehicle Classes") +
	scale_fill_tolkien(discrete=TRUE, option = 'Sauron') +
	ylab("") +
	xlab("Display")

ggsave("examples/hufflepuff_histogram.png", gg, width = 300, height = 110, units = "mm", device = "png", dpi = "retina")




# gryffindor_volcano ------------------------------------------------------

pal1 <- tolkien(8, house = "Sauron") # Left Image
gg1 <- ggplotify::as.ggplot(~image(volcano, col = pal1))


pal2 <- tolkien(128, house = "Sauron") # Right Image
gg2 <- ggplotify::as.ggplot(~image(volcano, col = pal2))

gg <- grid.arrange(gg1,gg2, ncol = 2)

ggsave("examples/sauron_volcano.png", gg, width = 300, height = 110, units = "mm", device = "png", dpi = "retina")




